File: 2461-amount-of-time-for-binary-tree-to-be-infected/amount-of-time-for-binary-tree-to-be-infected.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    Map<Integer, List<Integer>> graph = new HashMap<>();
    public int amountOfTime(TreeNode root, int start) {
        createGraph(root);
        Queue<Integer> q = new ArrayDeque<>();
        q.add(start);
        Set<Integer> visited = new HashSet<>();
        visited.add(start);
        int time = -1;
        while (!q.isEmpty()) {
            int size = q.size();
            for (int i = 0; i < size; i++) {
                int curr = q.poll();
                visited.add(curr);
                for (int neighbor : graph.get(curr)) {
                    if (visited.contains(neighbor)) continue;
                    q.add(neighbor);
                }
            }
            time++;
        }
        return time;
    }

    private void createGraph(TreeNode root) {
        graph.putIfAbsent(root.val, new ArrayList<>());
        if (root.left != null) {
            graph.get(root.val).add(root.left.val);
            graph.putIfAbsent(root.left.val, new ArrayList<>());
            graph.get(root.left.val).add(root.val);
            createGraph(root.left);
        }
        if (root.right != null) {
            graph.get(root.val).add(root.right.val);
            graph.putIfAbsent(root.right.val, new ArrayList<>());
            graph.get(root.right.val).add(root.val);
            createGraph(root.right);
        }
    }
}

File: 2461-amount-of-time-for-binary-tree-to-be-infected/amount-of-time-for-binary-tree-to-be-infected.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def amountOfTime(self, root: Optional[TreeNode], start: int) -> int:
        nodeToParent = defaultdict()
        q = deque()
        visited = set()
        def dfs(root):
            if not root:
                return
            if start == root.val:
                q.append(root)
                visited.add(root)
            if root.left:
                nodeToParent[root.left] = root
            if root.right:
                nodeToParent[root.right] = root
            dfs(root.left)
            dfs(root.right)
        dfs(root)
        days = 0
        while q:
            size = len(q)
            for i in range(size):
                curr = q.popleft()
                parent = nodeToParent[curr] if curr in nodeToParent else None
                if parent and parent not in visited:
                    q.append(parent)
                    visited.add(parent)
                if curr.left and curr.left not in visited:
                    q.append(curr.left)
                    visited.add(curr.left)
                if curr.right and curr.right not in visited:
                    q.append(curr.right)
                    visited.add(curr.right)
            days += 1
        return days - 1
                